File: src/odioctl/web/render.py
"""The page, built out of web/templates/*.html — no HTTP, no state changes.

Templates are `string.Template` (`$name` placeholders): whole sections as
well as the one-element partials (banner, hint, option, …), so the only
markup left in Python is the bare <h1> of the 403/404/405 dead ends in
`server`. The stylesheet and logo in web/static/ mirror odio-ui's look
(go-odio-api). Every value substituted into a template is HTML-escaped by
`escape` unless it is itself rendered HTML.

`render_page` reads from `Services` and never writes: the one exception is
`pop_action_result`, which is a read that consumes — the modal shows an
action's output once and the next page load is clean.
"""

from __future__ import annotations

import functools
import html
import importlib.resources
import socket
import string
from collections.abc import Callable
from typing import Any, cast

from odioctl import __version__, components, dac, state
from odioctl.upgrade import check
from odioctl.web.config import ODIO_UI_PORT
from odioctl.web.services import ActionResult, Services

_RES = importlib.resources.files("odioctl.web")

STATIC_FILES = {"style.css": "text/css; charset=utf-8", "logo.png": "image/png"}


@functools.cache
def _template(name: str) -> string.Template:
    return string.Template((_RES / "templates" / name).read_text(encoding="utf-8"))


@functools.cache
def static_asset(name: str) -> tuple[bytes, str] | None:
    """(content, media type) for a file under web/static/, or None if unknown."""
    if name not in STATIC_FILES:
        return None
    return (_RES / "static" / name).read_bytes(), STATIC_FILES[name]


def _render(template: str, **values: object) -> str:
    return _template(template).substitute(values)


def escape(value: object) -> str:
    """HTML-escape a value on its way into a template."""
    return html.escape(str(value), quote=True)


# Nearly every substituted value goes through it, hence the short local name.
_e = escape


# The sections receive a `render` with the per-process form token already bound
# rather than the token itself: every template carrying a form uses the same
# `$token` hidden field, so it is escaped once, here. Templates without the
# placeholder ignore the extra value.
Render = Callable[..., str]


def _renderer(token: str) -> Render:
    return functools.partial(_render, token=_e(token))


def _banner(kind: str, text: str) -> str:
    return _render("banner.html", kind=kind, text=_e(text)) if text else ""


def _hint(text: str) -> str:
    return _render("hint.html", text=_e(text))


def _section(title: str, body: str) -> str:
    """A section whose whole content is one banner or hint (the empty/unavailable states)."""
    return _render("section_banner.html", title=_e(title), banner=body)


# (chip text, button label) per component status; the button performs the opposite action.
_STATUS_UI = {
    "installed": ("Installed", "Disable"),
    "excluded": ("Disabled", "Enable"),
    "default": ("Will install on next upgrade", "Skip"),
}


ActionState = Callable[[str, str, str], tuple[str, str]]


def _component_actions(
    c: components.Component, render: Render, state_of: ActionState
) -> tuple[str, str]:
    """(buttons, notes) for one component's catalog actions.

    Offered only once the component is installed — the command it runs ships
    with the package. Each pending link (and the outcome of the last run) is
    rendered under the row, next to what it belongs to.
    """
    if not c.actions or c.status != "installed":
        return "", ""
    buttons: list[str] = []
    notes: list[str] = []
    for a in c.actions:
        buttons.append(
            render(
                "component_action_run.html",
                kind=_e(c.kind),
                name=_e(c.name),
                action=_e(a.id),
                button=_e(a.label),
            )
        )
        url, note = state_of(c.kind, c.name, a.id)
        if url:
            notes.append(
                render(
                    "component_action_link.html",
                    url=_e(url),
                    label=_e(a.link_label),
                    note=_e(a.link_note or "started"),
                )
            )
        elif note:
            notes.append(render("component_action_note.html", note=_e(f"{a.label}: {note}")))
    return "".join(buttons), "".join(notes)


def _component_row(
    c: components.Component, render: Render, child: bool, state_of: ActionState
) -> str:
    chip, button = _STATUS_UI[c.status]
    action = render(
        "component_action.html",
        kind=_e(c.kind),
        name=_e(c.name),
        enabled="0" if c.enabled else "1",
        button=button,
    )
    runs, notes = _component_actions(c, render, state_of)
    return render(
        "component_row.html",
        child=" child" if child else "",
        label=_e(c.label),
        description=_e(c.description or c.name),
        status=_e(c.status),
        chip=chip,
        action=action,
        runs=runs,
        notes=notes,
    )


def _components_section(
    st: state.State | None, err: str | None, render: Render, state_of: ActionState
) -> str:
    if st is None:
        return _section("Components", _banner("err", f"state.json: {err}"))
    comps = components.list_components(st)
    by_parent: dict[str, list[components.Component]] = {}
    orphans: list[components.Component] = []
    for f in comps:
        if f.kind != "feature":
            continue
        if f.parent:
            by_parent.setdefault(f.parent, []).append(f)
        else:
            orphans.append(f)
    rows_by_group: dict[str, list[str]] = {g: [] for g in components.GROUPS}
    infra: list[str] = []
    for r in comps:
        if r.kind != "role":
            continue
        if not r.toggleable:
            infra.append(r.label)
            continue
        rows = rows_by_group.setdefault(r.group, [])
        rows.append(_component_row(r, render, False, state_of))
        rows.extend(_component_row(f, render, True, state_of) for f in by_parent.get(r.name, []))
    rows_by_group[components.GROUPS[-1]].extend(
        _component_row(f, render, False, state_of) for f in orphans
    )
    groups = "".join(
        render("component_group.html", title=_e(title), rows="".join(rows))
        for title, rows in rows_by_group.items()
        if rows
    )
    # The install mode and target user drive what an upgrade actually installs,
    # so they belong here rather than in the page header.
    return render(
        "components.html",
        user=_e(st["target_user"]),
        mode=_e(st["install_mode"]),
        note=_e(components.APPLY_NOTE),
        groups=groups,
        infra=_e("Always installed: " + ", ".join(infra)) if infra else "",
    )


def _dac_section(d: dict[str, Any], render: Render) -> str:
    if not d["supported"]:
        return _section(
            "DAC",
            _banner(
                "warn", "No config.txt found — DAC selection is only available on Raspberry Pi."
            ),
        )
    current = d["current"]
    opts = [
        _render(
            "option.html",
            id="",
            attrs=" disabled selected" if not current else " disabled",
            text="— not configured —",
        )
    ]
    opts += [
        _render(
            "option.html",
            id=_e(e.id),
            attrs=" selected" if e.id == current else "",
            text=f"{_e(e.label)} ({_e(e.id)})",
        )
        for e in dac.CATALOG
    ]
    if current:
        cur = f"Current: {current}" + (
            " (managed by odioctl)" if d["managed"] else " (from existing config.txt)"
        )
    elif d["stray_lines"] and not d["managed"]:
        cur = "Unrecognised audio configuration in config.txt: " + "; ".join(d["stray_lines"])
    else:
        cur = "No DAC configured"
    unset = render("dac_unset.html") if d["managed"] else ""
    stray = ""
    if d["stray_lines"] and d["managed"]:
        # Unmanaged lines are simply what defines `current`; once odioctl owns the
        # block, anything else left active is a conflict worth flagging.
        stray = _banner(
            "warn",
            "Audio lines outside the odioctl block (commented out on Apply): "
            + "; ".join(d["stray_lines"]),
        )
    return render(
        "dac.html",
        options="".join(opts),
        unset=unset,
        current=_e(cur),
        stray=stray,
    )


def _upgrade_section(report: check.UpgradeReport | None, render: Render, ui_url: str) -> str:
    if report is None:
        return _section(
            "Upgrade", _banner("warn", "No upgrade check has run yet (odio-check-upgrade.timer).")
        )
    if not report["upgrade_available"]:
        return _section(
            "Upgrade",
            _hint(f"Up to date — odio {report['current']} (checked {report['checked_at']})."),
        )
    items: list[str] = []
    if report["latest"] != report["current"]:
        items.append(f"odio {_e(report['current'])} → {_e(report['latest'])}")
    items.extend(
        f"{_e(r['name'])} {_e(r['installed'])} → {_e(r['available'])}" for r in report["roles"]
    )
    for ref in report["pending_components"]:
        kind, _, name = ref.partition(":")
        items.append(f"install {_e(components.label_of(cast(components.Kind, kind), name))}")
    return render(
        "upgrade.html",
        items="".join(_render("upgrade_item.html", text=i) for i in items),
        ui_url=_e(ui_url),
    )


def _modal(result: ActionResult | None, render: Render) -> str:
    """The action's output, over the page. Plain HTML: no JS anywhere, so it is
    the POST response that carries it and `Close` is a link back to the page."""
    if result is None:
        return ""
    link = (
        render("modal_link.html", url=_e(result.url), label=_e(result.link_label))
        if result.url
        else ""
    )
    return render(
        "modal.html",
        title=_e(result.title),
        output=_e(result.output.strip() or "(no output)"),
        link=link,
    )


def render_page(services: Services, *, message: str = "", error: str = "", host: str = "") -> str:
    st, err = services.read_state()
    d = services.dac_status()
    render = _renderer(services.token)
    # The Host header when the browser gave one (that name reaches the box), the
    # box's own hostname otherwise — same address for the odio-ui link and ssh.
    # The logo is that way home: this page is a settings annex of odio-ui.
    hostname = host or socket.gethostname()
    ui_url = f"http://{hostname}:{ODIO_UI_PORT}/ui"
    version_badge = ""
    if st is not None:
        version_badge = _render("version_badge.html", odios=_e(st["odios"]))
    banners = [_banner("ok", message), _banner("err", error)]
    if d["reboot_required"]:
        banners.append(_banner("warn", "A reboot is required to apply the DAC change."))
    return render(
        "page.html",
        version=_e(__version__),
        hostname=_e(socket.gethostname()),
        ui_url=_e(ui_url),
        version_badge=version_badge,
        banners="".join(banners),
        components=_components_section(st, err, render, services.action_state),
        upgrade=_upgrade_section(services.upgrade_report(), render, ui_url),
        dac=_dac_section(d, render),
        modal=_modal(services.pop_action_result(), render),
    )

File: web/render.go
package web

// The page: view models built from the App, markup and composition in
// templates/*.gohtml, look and htmx copies from odio-ui in static/. Each
// section swaps itself on its own /events event; a POST answers the notice.

import (
	"embed"
	"fmt"
	"html/template"
	"os"
	"strings"

	"github.com/b0bbywan/odioctl/components"
	"github.com/b0bbywan/odioctl/config"
	"github.com/b0bbywan/odioctl/dac"
	"github.com/b0bbywan/odioctl/state"
	"github.com/b0bbywan/odioctl/upgrade"
)

//go:embed templates/*.gohtml
var templatesFS embed.FS

//go:embed static/style.css static/logo.png static/htmx.min.js static/htmx-sse.js
var staticFS embed.FS

var templates = template.Must(template.ParseFS(templatesFS, "templates/*.gohtml"))

var staticTypes = map[string]string{
	"style.css":   "text/css; charset=utf-8",
	"logo.png":    "image/png",
	"htmx.min.js": "text/javascript; charset=utf-8",
	"htmx-sse.js": "text/javascript; charset=utf-8",
}

// StaticAsset is the (content, media type) of a file under static/, or ok=false.
func StaticAsset(name string) (content []byte, mediaType string, ok bool) {
	ctype, known := staticTypes[name]
	if !known {
		return nil, "", false
	}
	b, err := staticFS.ReadFile("static/" + name)
	if err != nil {
		return nil, "", false
	}
	return b, ctype, true
}

// -- view models ---------------------------------------------------------

type bannerView struct{ Kind, Text string }

type actionView struct {
	ID, Button               string
	URL, LinkLabel, LinkNote string     // pending link, when URL is set
	Note                     actionNote // how the last run ended: a badge, or the failure in red
}

type rowView struct {
	Child                     bool
	Label, Description        string
	Status                    string // picks the icon pair, and its colours
	Title                     string // the state and what a click does: title/aria-label
	Required                  bool   // a lock, where the toggle would be
	Token, Kind, Name, Enable string // the toggle form
	Actions                   []actionView
}

type groupView struct {
	Title string
	Rows  []rowView
}

type componentsView struct {
	Err    string // non-empty → error banner instead of content
	Groups []groupView
}

type dacOptionView struct {
	ID, Text           string
	Disabled, Selected bool
}

type dacView struct {
	Supported      bool
	Token, Current string
	Managed        bool
	Options        []dacOptionView
	Stray          string
}

type upgradeView struct {
	Checked   bool // a check has run (upgrades.json exists)
	Available bool
	Running   bool       // odio-upgrade.service is being followed: no Apply button
	Note      actionNote // how the last run ended, empty when none was seen
	UpToDate  string
	Token     string
	Items     []string
}

type pageView struct {
	Version, Base, UIURL, Hostname string
	Odios                          string          // "" = no badge
	Sections                       []template.HTML // each rendered by its own template, in sections order
}

// noticeView is the outcome of a POST: its banner, and the modal of an
// action — the fragment for #notice, with the modal swapped out of band
// into #modal.
type noticeView struct {
	Notices []bannerView
	Modal   *ActionResult
}

// The icon carries no words: its title says the state and what a click does.
// No verb: a label can be plural ("Web radios").
var statusTitle = map[components.Status]string{
	components.Installed: "%s — installed, click to disable",
	components.Excluded:  "%s — disabled, click to enable",
	components.Default:   "%s — installs on the next upgrade, click to skip",
}

func rowViewOf(app *App, c components.Component, child bool) rowView {
	title := fmt.Sprintf(statusTitle[c.Status], c.Label)
	if !c.Toggleable {
		title = c.Label + " — required by odio"
	}
	enable := "1"
	if c.Enabled() {
		enable = "0"
	}
	row := rowView{
		Child:       child,
		Label:       c.Label,
		Description: c.Description,
		Status:      string(c.Status),
		Title:       title,
		Required:    !c.Toggleable,
		Token:       app.Token(),
		Kind:        string(c.Kind),
		Name:        c.Name,
		Enable:      enable,
	}
	// Actions are offered only once the component is installed — the command
	// they run ships with the package. The pending link (or the outcome of
	// the last run) renders under the row, next to what it belongs to.
	if c.Status != components.Installed {
		return row
	}
	for _, a := range c.Actions {
		url, note := app.actions.State(c.Kind, c.Name, a.ID)
		av := actionView{ID: a.ID, Button: a.Label, Note: note}
		if url != "" {
			av.URL, av.LinkLabel, av.LinkNote = url, a.LinkLabel, a.LinkNote
		}
		row.Actions = append(row.Actions, av)
	}
	return row
}

func componentsViewOf(app *App, st *state.State, stateErr string) componentsView {
	if st == nil {
		return componentsView{Err: "state.json: " + stateErr}
	}
	comps := components.List(*st, app.TargetManifest())
	byParent := map[string][]components.Component{}
	var orphans []components.Component
	for _, f := range comps {
		if f.Kind != components.Feature {
			continue
		}
		if f.Parent != "" {
			byParent[f.Parent] = append(byParent[f.Parent], f)
		} else {
			orphans = append(orphans, f)
		}
	}
	rowsByGroup := map[string][]rowView{}
	for _, r := range comps {
		if r.Kind != components.Role {
			continue
		}
		rows := append(rowsByGroup[r.Group], rowViewOf(app, r, false))
		for _, f := range byParent[r.Name] {
			rows = append(rows, rowViewOf(app, f, true))
		}
		rowsByGroup[r.Group] = rows
	}
	last := components.Groups[len(components.Groups)-1]
	for _, f := range orphans {
		rowsByGroup[last] = append(rowsByGroup[last], rowViewOf(app, f, false))
	}
	var view componentsView
	for _, title := range components.Groups {
		if rows := rowsByGroup[title]; len(rows) > 0 {
			view.Groups = append(view.Groups, groupView{Title: title, Rows: rows})
		}
	}
	return view
}

func dacViewOf(app *App, d dac.Status) dacView {
	if !d.Supported {
		return dacView{}
	}
	view := dacView{
		Supported: true,
		Token:     app.Token(),
		Managed:   d.Managed,
		Options: []dacOptionView{
			{ID: "", Text: "— not configured —", Disabled: true, Selected: d.Current == ""},
		},
	}
	for _, e := range dac.Catalog {
		view.Options = append(view.Options, dacOptionView{
			ID:       e.ID,
			Text:     fmt.Sprintf("%s (%s)", e.Label, e.ID),
			Selected: e.ID == d.Current,
		})
	}
	switch {
	case d.Current != "" && d.Managed:
		view.Current = "Current: " + d.Current + " (managed by odioctl)"
	case d.Current != "":
		view.Current = "Current: " + d.Current + " (from existing config.txt)"
	case len(d.StrayLines) > 0:
		view.Current = "Unrecognised audio configuration in config.txt: " + strings.Join(d.StrayLines, "; ")
	default:
		view.Current = "No DAC configured"
	}
	if len(d.StrayLines) > 0 && d.Managed {
		// Unmanaged lines are simply what defines Current; once odioctl owns
		// the block, anything else left active is a conflict worth flagging.
		view.Stray = "Audio lines outside the odioctl block (commented out on Apply): " +
			strings.Join(d.StrayLines, "; ")
	}
	return view
}

func upgradeViewOf(app *App, report *upgrade.Report) upgradeView {
	running, note := app.upgrades.State(report)
	view := upgradeView{Running: running, Note: note}
	if report == nil {
		return view
	}
	view.Checked = true
	view.Available = report.UpgradeAvailable
	view.Token = app.Token()
	if !report.UpgradeAvailable {
		view.UpToDate = fmt.Sprintf("Up to date — odio %s (checked %s).",
			report.Current, report.CheckedAt)
		return view
	}
	if report.Latest != report.Current {
		view.Items = append(view.Items, fmt.Sprintf("odio %s → %s", report.Current, report.Latest))
	}
	for _, r := range report.Roles {
		view.Items = append(view.Items, fmt.Sprintf("%s %s → %s", r.Name, r.Installed, r.Available))
	}
	for _, ref := range report.PendingComponents {
		kind, name, _ := strings.Cut(ref, ":")
		view.Items = append(view.Items, "install "+components.LabelOf(&report.Manifest, components.Kind(kind), name))
	}
	return view
}

func modalView(res *ActionResult) *ActionResult {
	if res == nil {
		return nil
	}
	m := *res
	m.Output = strings.TrimSpace(m.Output)
	if m.Output == "" {
		m.Output = "(no output)"
	}
	return &m
}

func noticeViewOf(msg, errText string, modal *ActionResult) noticeView {
	view := noticeView{Modal: modalView(modal)}
	for _, b := range []bannerView{{"ok", msg}, {"err", errText}} {
		if b.Text != "" {
			view.Notices = append(view.Notices, b)
		}
	}
	return view
}

// bannersView is what holds across renders: the reboot the DAC change
// waits for, with the button that does it, and the refusal of the last one.
type bannersView struct {
	RebootRequired bool
	RebootError    string
	Token          string
}

func bannersOf(app *App, d dac.Status) bannersView {
	return bannersView{RebootRequired: d.RebootRequired, RebootError: app.RebootError(), Token: app.Token()}
}

// RenderNotice is the answer to a POST: the banner for #notice and, out of
// band, the modal of an action. The state follows on the stream.
func RenderNotice(msg, errText string, modal *ActionResult) (string, error) {
	var b strings.Builder
	if err := templates.ExecuteTemplate(&b, "notice.gohtml", noticeViewOf(msg, errText, modal)); err != nil {
		return "", err
	}
	return b.String(), nil
}

// Section is one piece of the page and one event on the stream: its
// template, <Name>.gohtml, has a root listening to Name (sse-swap) that
// swaps itself. sections is the page's order.
type Section struct {
	Name string
	view func(app *App) any
}

var sections = []Section{
	{"banners", func(app *App) any { return bannersOf(app, app.DacStatus()) }},
	{"upgrade", func(app *App) any { return upgradeViewOf(app, app.UpgradeReport()) }},
	{"components", func(app *App) any { st, err := stateOf(app); return componentsViewOf(app, st, err) }},
	{"dac", func(app *App) any { return dacViewOf(app, app.DacStatus()) }},
}

// SectionNames is every section, what a stream sends first.
func SectionNames() []string {
	names := make([]string, len(sections))
	for i, sec := range sections {
		names[i] = sec.Name
	}
	return names
}

func render(name string, data any) (string, error) {
	var b strings.Builder
	err := templates.ExecuteTemplate(&b, name, data)
	return b.String(), err
}

// Fragment is one event on the stream.
type Fragment struct {
	Event, HTML string
}

// RenderFragments is what a change named: sections in the page's order,
// then the modal of a finished action by its id (only a page showing it
// listens). A name that is neither is nothing.
func RenderFragments(app *App, names []string) ([]Fragment, error) {
	want := map[string]bool{}
	for _, n := range names {
		want[n] = true
	}
	var out []Fragment
	for _, sec := range sections {
		if !want[sec.Name] {
			continue
		}
		html, err := render(sec.Name+".gohtml", sec.view(app))
		if err != nil {
			return nil, err
		}
		out = append(out, Fragment{sec.Name, html})
	}
	for _, res := range app.actions.Finished() {
		if !want[res.ID] {
			continue
		}
		html, err := render("modal.gohtml", modalView(res))
		if err != nil {
			return nil, err
		}
		out = append(out, Fragment{res.ID, html})
	}
	return out, nil
}

// stateOf is state.json as the Components section takes it: the state, or
// nil and the message to show instead.
func stateOf(app *App) (*state.State, string) {
	s, err := app.ReadState()
	if err != nil {
		return nil, stateErrorMsg(app.Config().StatePath, err)
	}
	return &s, ""
}

// RenderPage is GET /: the sections as they stand, an empty #notice and
// #modal for the POSTs to fill, every URL relative to the <base> o sets.
func RenderPage(app *App, o origin) (string, error) {
	// The Host header when the browser gave one (that name reaches odio), its
	// own hostname otherwise — same address for the odio-ui link and ssh.
	hostname := o.Host
	if hostname == "" {
		hostname, _ = os.Hostname()
	}
	uiURL := app.Config().UIURL
	if uiURL == "" {
		uiURL = fmt.Sprintf("http://%s:%d/ui", hostname, OdioUIPort)
	}
	selfName, _ := os.Hostname()

	view := pageView{
		Version:  config.AppVersion,
		Base:     o.Base,
		UIURL:    uiURL,
		Hostname: selfName,
	}
	if st, _ := stateOf(app); st != nil {
		view.Odios = st.Odios
	}
	for _, sec := range sections {
		html, err := render(sec.Name+".gohtml", sec.view(app))
		if err != nil {
			return "", err
		}
		view.Sections = append(view.Sections, template.HTML(html)) // our own template's output, already escaped
	}

	var b strings.Builder
	if err := templates.ExecuteTemplate(&b, "page.gohtml", view); err != nil {
		return "", err
	}
	return b.String(), nil
}
